packages/synchronizer/src/__tests__/projectSynchronizer.spec.ts

```diff
@@ -0,0 +1,252 @@
+    }
+
+    await cleanupTmpConfigDir();
+  });
+
+  it('returns empty data if not synchronized', async () => {
+    const storagePath = await createTmpConfigDir();
+    const synchronizer = createSynchronizer(storagePath, stubs);
+
+    const info = synchronizer.getProjectInfo(storagePath);
+    assert.isUndefined(info);
+
+    const permissions = synchronizer.getProjectPermissions(storagePath);
+    assert.isUndefined(permissions);
+
+    const policy = synchronizer.getProjectPolicy(storagePath);
+    assert.isObject(policy);
+    assert.isFalse(policy.valid);
+    assert.isEmpty(policy.path);
+    assert.isEmpty(policy.policy);
+
+    const suppressions = synchronizer.getRepositorySuppressions(storagePath);
+    assert.isArray(suppressions);
+    assert.isEmpty(suppressions);
+  });
+
+  it('throws error when no access token provided for fetch', async () => {
+    try {
+      const storagePath = await createTmpConfigDir();
+      const synchronizer = createSynchronizer(storagePath, stubs);
+
+      await synchronizer.synchronize({} as any, storagePath);
+
+      assert.fail('Should have thrown error.');
+    } catch (err: any) {
+      assert.match(err.message, /Cannot fetch data without access token/);
+    }
+  });
+
+  it('returns valid data after synchronization (repo path)', async () => {
+    const storagePath = await createTmpConfigDir();
+    const synchronizer = createSynchronizer(storagePath, stubs);
+
+    const queryApiStub = sinon.stub((synchronizer as any)._apiHandler, 'queryApi').callsFake(async (...args) => {
+      const query = args[0] as string;
+
+      if (query.includes('query getUser')) {
+        return {
+          data: {
+            me: {
+              id: 1,
+              email: 'user1@kubeshop.io',
+              projects: [
+                {
+                  project: {
+                    id: 1000,
+                    slug: 'user1-proj',
+                    name: 'User1 Project',
+                    repositories: [
+                      {
+                        id: 'user1-proj-policy-id',
+                        projectId: 1000,
+                        provider: 'GITHUB',
+                        owner: 'kubeshop',
+                        name: 'monokle-core',
+                        prChecks: false,
+                        canEnablePrChecks: true,
+                      },
+                    ],
+                  },
+                },
+              ],
+            },
+          },
+        };
+      }
+
+      if (query.includes('query getProject')) {
+        return {
+          data: {
+            getProject: {
+              id: 1000,
+              slug: 'user1-proj',
+              name: 'User1 Project',
+              projectRepository: {
+                id: 'repo1',
+                projectId: 1000,
+                provider: 'GITHUB',
+                owner: 'kubeshop',
+                name: 'monokle-demo',
+                suppressions: [{
```

Comment:
I apologise since we talked about one big query beforehand.

It just comes to mind that this large query would both have high complexity and be frequently queried. That makes this an idea that does not scale well. Furthermore, policies, repositories and suppressions rarely change.

I wonder if we should move towards an offline-first approach with a SQLite database and this main, frequently polled query simply contains `updatedAt` / `content-hash` / `lastCursor` for policy / repositories / suppressions. That makes the frequently polled query lightweight, after which heavier synchronises can happen when desynced.

Given that there is just one repository and the policy is quite small, we could get away with only taking this road for suppressions. Alternatively, it could also just be `repository.suppressions.json` for now to keep it simple, stupid.

Reply:
We could also use the `getSuppressions` query (instead of the `suppressions` field on the repository model) which accepts `.from` and `.to` arguments. This would work well with the cache implementation described by @WitoDelnat. We should also implement proper pagination (with [grapqhl connections](https://relay.dev/graphql/connections.htm)) so that we can paginate the first query. The basis for implementing pagination with connections is already [in place](https://github.com/kubeshop/monokle-saas/blob/main/apiDomain/apiService/src/utils/pagination.ts)

Reply:
> which accepts `.from` and `.to` arguments.

This is interesting 👍 So with this, in fact, we could have incremental updates to stored suppressions, right? To fetch changes for suppressions only from most recent fetch.

> We should also implement proper pagination (with [grapqhl connections](https://relay.dev/graphql/connections.htm)) so that we can paginate the first query. The basis for implementing pagination with connections is already [in place](https://github.com/kubeshop/monokle-saas/blob/main/apiDomain/apiService/src/utils/pagination.ts)

Not sure how using pagination could help here? 🤔 We still need entire suppressions list for validation. Could you expand on that?

---

From what I checked, looking into current db schema we also have `updatedAt` field for policy (`ProjectPolicy`) so this can be used as @WitoDelnat suggested to not pull entire policy contents every time.